### PR TITLE
Tabby: bump 0.7.0 -> 0.8.3 and add systemd service

### PR DESCRIPTION
**Pull request by ghthor**
## Description of changes

1. followup to #291744 
1. Upgraded tabby to release 0.8.3
2. Replaced static linking of vendored llama-cpp with dynamic link to llama-cpp from nixpkgs
3. Add systemd service for [tabby, tabby-scheduler] with timer configuration for tabby scheduler

### Upgraded to upstream 0.8.3 release

Added support for rocm AMD acceleration.

- [x] tested rocm acceleration locally with AMD Radeon RX 560 Series
- [x] tested cuda acceleration locally with NVIDIA Geforce RTX 2080 SUPER
- [x] tested darwin metal on an M1 macbook

Both tests used the new systemd service to run tabby

### Dynamic Linking w/ llama-cpp

I was having compile errors with the vendored copy of llama-cpp that tabbyml was statically compiling, so I switched to dynamic linking with the llama-cpp that is within nixpkgs.

### Systemd service

Example of the configuration required here.

https://github.com/ghthor/shrc/blob/18109599c526e580e54b11a16b6c1ef017777918/nix/cryptnix/configuration.nix#L434

```nix
  services.tabby = {
    enable = true;
    acceleration = "cuda";
    model = "TabbyML/StarCoder-3B";

    settings = {
      repositories = [
        {
          # Has to be created/updated manually, and permissions corrected
          git_url = "file:///var/lib/tabby/source_repo";
        }
      ];
    };
  };
```

Using the service makes it easy to enable a specific type of acceleration.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ghthor** [comment]
@sersorrel I was able to test the rocm support on my local workstation, still pinging you if you'd like to test as well.

**Comment by ghthor** [comment]
@happysalada ready for review

File: pkgs/by-name/ta/tabby/package.nix [hunk 1]
```diff
@@ -1,36 +1,76 @@
...
 , pkg-config
 , protobuf
-, rustPlatform
-, addOpenGLRunpath
-, cudatoolkit
-, nvidia ? true
+
+, llama-cpp
+# one of `[ null "rocm" "cuda" ]`
+, acceleration ? null
```
Comment (happysalada) [inline pkgs/by-name/ta/tabby/package.nix:14]: I would either
- not have that configurable and just use config.cudaSupport and rocmsupport
- or at least have the defaults being picked up by the config.

**Review by happysalada — COMMENTED** [review]
This is looking good, thank you !
I have a couple of questions / comments

**Review by keysmashes — COMMENTED** [review]
Looks good! I built with `nix-build -E 'with import ./. {}; tabby.override { acceleration = "rocm"; }'`, and `./result/bin/tabby serve --model TabbyML/StarCoder-1B --device rocm` appeared to work fine.

What's the purpose of the `modelDownload` config? I didn't explicitly run `tabby download`, it just downloaded it automatically.

File: nixos/modules/services/misc/tabby.nix [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+          up. This will also have the effect of performing model updates
+          whenever tabby is started.
+
+          A model can be manually downloaded leaving this option false with a
+          command like the following
+
+          > TABBY_ROOT=/var/lib/tabby tabby download --model TabbyML/StarCoder-1B
+        '';
+      };
+      acceleration = lib.mkOption {
```
[the remark this answers is not in the document]
Comment (ghthor) [reply nixos/modules/services/misc/tabby.nix:51]: Yep, I think that makes sense, will fix this

File: nixos/modules/services/misc/tabby.nix [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        StateDirectory = [ "tabby" ];
+        ConfigurationDirectory = [ "tabby" ];
+        DynamicUser = true;
+        User = "tabby";
+        Group = "tabby";
+      };
+
+      serviceEnv = lib.mkMerge [
+        {
+          TABBY_ROOT = "%S/tabby";
```
[the remark this answers is not in the document]
Comment (ghthor) [reply nixos/modules/services/misc/tabby.nix:136]: Screenshot_2024-03-03_22-03-48

https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html

I think `%S` just resolves to `/var/lib`.

Used this as reference https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/ollama.nix#L48

And confirmed that `TABBY_ROOT=/var/lib/tabby` after systemd fills in `%S` with the following command

```
❯ sudo cat /proc/$(systemctl show --property MainPID tabby | cut -d= -f2)/environ | xargs --null --max-args=1 echo | grep ^TABBY
TABBY_DISABLE_USAGE_COLLECTION=1
TABBY_ROOT=/var/lib/tabby
```

File: nixos/modules/services/misc/tabby.nix [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    services.tabby = {
+      enable = lib.mkEnableOption (
+        lib.mdDoc "Server for local large language models"
+      );
+
+      package = lib.mkPackageOption pkgs "tabby" { };
+
+      port = lib.mkOption {
+        type = types.port;
+        default = 11434;
```
[the remark this answers is not in the document]
Comment (ghthor) [reply nixos/modules/services/misc/tabby.nix:22]: It is, I didn't change it from the copy I made of ollama that I started from. I'll pick a different port

**Comment by ghthor** [comment]
> What's the purpose of the modelDownload config? I didn't explicitly run tabby download, it just downloaded it automatically.

Yep, I can remove this extra configuration. I started using tabby before June 1, 2023 when they added this commit[1], didn't realize

[1] https://github.com/TabbyML/tabby/commit/ca077a34039a9e7bfe0de036802772d673b4b4e7

File: nixos/modules/services/misc/tabby.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (ghthor) [reply nixos/modules/services/misc/tabby.nix:39]: I added some defaulting of the acceleration type in the presence of either `rocmSupport or cudaSupport` on `nixpkgs.config`. Even with that defaulting I have left `acceleration` as a service and package configuration value to handle the case when both of the above are present, as we can't make a clear decision and we do have to pick a single option. See this comment I've added to the code where the `llama-cpp` package is overriden.

https://github.com/NixOS/nixpkgs/pull/292873/commits/274d9b9ed96d4b32fa5173827814daaac914fa6e#diff-abdbe2b562428d2ed485d33ee478cf9e126faae74d5cc60b48284c7c254ddf33R86

File: pkgs/by-name/ta/tabby/package.nix (hunk 1 above)
Comment (ghthor) [reply pkgs/by-name/ta/tabby/package.nix]: See comment in other thread

https://github.com/NixOS/nixpkgs/pull/292873#discussion_r1510808797

File: nixos/modules/services/misc/tabby.nix [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    };
+
+
+    systemd = let
+      serviceUser = {
+        WorkingDirectory = "/var/lib/tabby";
+        StateDirectory = [ "tabby" ];
+        ConfigurationDirectory = [ "tabby" ];
+        DynamicUser = true;
+        User = "tabby";
```
[the remark this answers is not in the document]
Comment (ghthor) [reply nixos/modules/services/misc/tabby.nix:152]: https://github.com/NixOS/nixpkgs/pull/292873/commits/7cc929f28a2f1d162b92ea6ad0777ab37bd88f25

**Comment by happysalada** [comment]
This looks good to me, happy to leave this open 1 more day to give some more time to other people to take a look.

File: pkgs/by-name/ta/tabby/package.nix [hunk 6]
```diff
@@ -1,36 +1,141 @@
...
+    (optional cudaSupport "cuda")
+    (optional rocmSupport "rocm")
+    (optional metalSupport "metal")
+  ];
+
+  warnIfMultipleAccelerationMethods = configured: (let
+    len = builtins.length configured;
+    result = if len == 0 then "cpu" else (builtins.head configured);
+  in
+    # lib.warnIf (len > 1) "multiple acceleration methods available: ${toString availableAccelerations}" "cpu"
```
Comment (de11n) [inline pkgs/by-name/ta/tabby/package.nix:46]: Is this comment meant to still be here?

File: pkgs/by-name/ta/tabby/package.nix [hunk 7]
```diff
@@ -1,36 +1,141 @@
...
+
+      To configure tabby with a specific acceleration method either
+
+      1. configure a package override
+      2. set `services.tabby.acceleration`
+
+      For example:
+        services.tabby.acceleration = "cuda";
+      Or:
+        services.tabby.acceleration = "rocm";
```
Comment (de11n) [inline pkgs/by-name/ta/tabby/package.nix:59]: It feels backward to mention NixOS module configuration in the warning of a specific package. Perhaps this validation deserves to be in the module instead. A shorter, less-verbose warning could be here instead.

**Review by de11n — COMMENTED** [review]
Extremely well done! A few small comments.

File: nixos/modules/services/misc/tabby.nix [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+          downloaded you'll need to move it into tabby's state directory which
+          lives in `/var/lib/tabby`. Because the tabby.service is configured to
+          use a DyanmicUser the service will need to have been started at least
+          once before you can move the locally existing model into
+          `/var/lib/tabby`. You can set the model to 'none' and tabby will
+          startup and fail to download a model, but will have created the
+          `/var/lib/tabby` directory. You can then copy over the model manually
+          into `/var/lib/tabby`, update the model option to the name you just
+          downloaded and copied over then `nixos-rebuild switch` to start using
+          it.
```
[the remark this answers is not in the document]
Comment (ghthor) [reply nixos/modules/services/misc/tabby.nix:46]: Yeah, I had a hard time writing that out and it isn't a simple nor straighforward process to explain due to the implementations details of `DynamicUser` systemd config leaking. I'd be happy to revisit this and make it simpler to load models that the user already downloaded outside of the scope of the systemd service in a followup.

File: nixos/modules/services/misc/tabby.nix [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+          -  `rocm`: supported by modern AMD GPUs
+          -  `cuda`: supported by modern NVIDIA GPUs
+          - `metal`: supported on darwin aarch64 machines
+
+          Tabby will try and determine what type of acceleration that is
+          already enabled in your configuration when `acceleration = null`.
+
+          - nixpkgs.config.cudaSupport
+          - nixpkgs.config.rocmSupport
+          - if stdenv.isDarwin && stdenv.isAarch64
```
[the remark this answers is not in the document]
Comment (ghthor) [reply nixos/modules/services/misc/tabby.nix:76]: It is supposed to convey where the package is auto configuring acceleration from in an attempt to help the user understand that in the face of multiple of these being set we're not able to make a good decision about which acceleration mode to use, thus suggest the user explicitly setting the acceleration method.

File: pkgs/by-name/ta/tabby/package.nix (hunk 6 above)
Comment (ghthor) [reply pkgs/by-name/ta/tabby/package.nix:46]: Nope, will remove

File: pkgs/by-name/ta/tabby/package.nix (hunk 7 above)
Comment (ghthor) [reply pkgs/by-name/ta/tabby/package.nix:59]: Yeah that makes sense, I'll move this around

File: nixos/modules/services/misc/tabby.nix (hunk 8 above)
[the remark this answers is not in the document]
Comment (de11n) [reply nixos/modules/services/misc/tabby.nix:46]: I wouldn't consider it blocking. That's fine.

File: pkgs/by-name/ta/tabby/package.nix (hunk 7 above)
Comment (de11n) [reply pkgs/by-name/ta/tabby/package.nix]: Looks better. Thanks. It looks like you have a trailing whitespace somewhere, causing the CI to fail.

**Comment by happysalada** [comment]
pkgs/by-name/ta/tabby/package.nix:
	70: Trailing whitespace

**Comment by happysalada** [comment]
I just noticed the darwin x86_64 build is broken. We should at least mark it broken for that.
If you dont know about aarch64 i would mark it broken as well. If youve tested on darwin aarch64 then no problem leaving it as is.

**Comment by ghthor** [comment]
Alright, little searching I found this https://github.com/NixOS/nixpkgs/pull/140325 for reference on how to mark the package as broken which I think points to it being a boolean flag in the meta object.

I don't have hardware to test linux-aarch64, but I will soon as that's what I'd like to try and run this on at work.

**Comment by ghthor** [comment]
Fixed commits to follow https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions

**Comment by happysalada** [comment]
Perfect thanks ! I plan on merging this when i go home today.

**Review by de11n — APPROVED** [review]

**Pull request merged**